core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1840,6 +1841,13 @@ private void handleErrorInSyncContext(Status error) {
       if (NameResolverListener.this.helper != ManagedChannelImpl.this.lbHelper) {
         return;
       }
+      // Apply Default Service Config if initial name resolution fails.
+      if (lastServiceConfig == EMPTY_SERVICE_CONFIG && defaultServiceConfig != null) {
```

Comment:
Yes, we initialize things to EMPTY_SERVICE_CONFIG, but we also use it after the initial resolution. And `defaultServiceConfig != null` shouldn't be relevant.

When we receive addresses but an invalid service config, we use do check whether there is a `defaultServiceConfig`. But if `serverConfigError == null`, then we still use EMPTY_SERVICE_CONFIG.

https://github.com/grpc/grpc-java/blob/2fe1a13cd07c86937d02c508025616c309424c9a/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L1748-L1772

If `serverConfigError != null` then we go through `onError()` which triggers `realChannel.onConfigError()` just a few lines up in this method. `onConfigError()` however doesn't check `defaultServiceConfig`:
https://github.com/grpc/grpc-java/blob/2fe1a13cd07c86937d02c508025616c309424c9a/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L957-L961

So I think the problem is "addresses + serviceConfigError" handles the default properly. But "address resolution error" ignores the `defaultServiceConfig` in `RealChannel.onConfigError()`. Looks like we should change `onConfigError()`, which is already doing proper checking to see if we've applied the service config already.

Reply:
@ejona86 @kannanjgithub i understand 2 below points from your above comments, please confirm the same.

Point 1 :  you mentioned in the below point the check **defaultServiceConfig != null**  shouldn't be relevant as it will have default value so this check is not required? but i feel adding additional null check will not harm anything as per the standards.
   if (lastServiceConfig == EMPTY_SERVICE_CONFIG && defaultServiceConfig != null) 

Point 2 : "address resolution error" ignores the defaultServiceConfig in RealChannel.onConfigError() ?  needs to updated as below ?

    void onConfigError() {
      if (configSelector.get() == INITIAL_PENDING_SELECTOR && **defaultServiceConfig == null)** {
        updateConfigSelector(null);
      }
    }

Reply:
So we already have this handling in NameResolverListener::onError (for address resolution errors) that ends up updating the config selector but with null.

 NameResolverListener::onError -> NameResolverListener::handleErrorInSyncContext -> realChannel.onConfigError -> 
  
      if (configSelector.get() == INITIAL_PENDING_SELECTOR) {
        updateConfigSelector(null);
      }

So the call to update  the config selector _already exists_ (that we missed noticing), only it uses null always and doesn't use the _defaultServiceConfig_. If we just change here to 
   updateConfigSelector(defaultServiceConfig); 
it would do. The check if (configSelector.get() == INITIAL_PENDING_SELECTOR) also already makes sure that the policy is only applied once, not on repeat invocations.

This means we didn't have to change anything in handleErrorInSyncContext. The other point was about the checks in that change for EMPTY_SERVICE_CONFIG was not correct but this value also gets assigned during a successful name resolution but with empty config, so our use of this value to mean that the config was never applied before was not correct. We need not continue discussion on this point since it turns out the code change in this method was not necessary to do.

Reply:
@kannanjgithub As discussed in the call changes are in progress.

Reply:
@kannanjgithub @ejona86 code changes are done for the recent review points. please have a look once.
also updated the UT as per my current understanding which is covered as part of onError() flow as I don't see in the existing cases where they are covering the onConfigError() directly  as its private class.

Reply:
There is ManagedChannelImpl::getConfigSelector() method that exposes its real channel's config selector. We should use this method and assert that it is same as the default service config. See example uses of this method in tests [here](https://github.com/grpc/grpc-java/blob/c63e3548835e838ded2f9eaf3be03dcb6b2b53a7/core/src/test/java/io/grpc/internal/ServiceConfigErrorHandlingTest.java#L471). 
Asserting for the config selector getting selected is important, asserting for channel trace events is secondary.

Reply:
@kannanjgithub Review point fixed and Request you have a look once.
unable to cover line lastServiceConfig = defaultServiceConfig as its final class so unable to mock this behavior also I have seen this in the existing cases as no where directly called this its covered as part of onError() flow.